core/src/main/java/org/infinispan/container/impl/BoundedSegmentedDataContainer.java

```diff
@@ -0,0 +1,253 @@
+package org.infinispan.container.impl;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.PrimitiveIterator;
+import java.util.Spliterator;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+import org.infinispan.commons.util.ComposeArraySpliterator;
+import org.infinispan.commons.util.ConcatIterator;
+import org.infinispan.commons.util.EntrySizeCalculator;
+import org.infinispan.commons.util.IntSet;
+import org.infinispan.container.entries.CacheEntrySizeCalculator;
+import org.infinispan.container.entries.InternalCacheEntry;
+import org.infinispan.container.entries.PrimitiveEntrySizeCalculator;
+import org.infinispan.eviction.EvictionType;
+import org.infinispan.marshall.core.WrappedByteArraySizeCalculator;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.CacheWriter;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.Policy;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+
+/**
+ * Bounded implementation of segmented data container. Bulk operations (iterator|spliterator) that are given segments
+ * use the segments maps directly to only read the given segments (non segment based just read from bounded container).
+ * <p>
+ * Note this implementation supports both temporary non owned segments and not (L1).
+ * @author wburns
+ * @since 9.3
+ */
+public class BoundedSegmentedDataContainer<K, V> extends DefaultSegmentedDataContainer<K, V> {
+   protected final Cache<K, InternalCacheEntry<K, V>> evictionCache;
+   protected final ConcurrentMap<K, InternalCacheEntry<K, V>> entries;
```

Comment:
I'm not sure why you have both `evictionCache` and `entries`, and you also extend `DefaultSegmentedDataContainer`

Reply:
The `evictionCache` is required soley to get at the eviction information.
The `entries` is the actual Map we write to for caffeine.
None of these deal with segmented information, the segmented data is still all stored in the maps in the `DefaultSegmentedDataContainer` for iteration and peek purposes.

When we write to the caffeine map we propagate the updates to the underlying segmented containers (essentially indexing that data).

Reply:
Ah ok, I thought the `entries` field was new, I didn't realize you had it from the start